File: client-control-m.js
const axios = require('axios');
const FormData = require('form-data');

/**
 * Cliente para ejecutar la API de Control-M desde la máquina local
 * @param {Object} controlMInfo - Información obtenida de la API principal
 * @returns {Promise<Object>} Resultado de la ejecución de Control-M
 */
async function executeControlMFromClient(controlMInfo) {
    try {
        console.log(`Ejecutando Control-M desde cliente: ${controlMInfo.url}`);

        // Convertir JSON a string y crear buffer
        const jsonString = JSON.stringify(controlMInfo.jsonData, null, 2);
        const jsonBuffer = Buffer.from(jsonString, 'utf8');

        // Crear form-data con el buffer
        const form = new FormData();
        form.append('definitionsFile', jsonBuffer, {
            filename: controlMInfo.filename,
            contentType: 'application/json'
        });

        // Configurar headers
        const config = {
            headers: {
                ...form.getHeaders(),
                ...controlMInfo.headers
            },
            timeout: 30000 // 30 segundos timeout
        };

        // Realizar la petición POST
        const response = await axios.post(controlMInfo.url, form, config);
        
        console.log(`Control-M ejecutado exitosamente. Status: ${response.status}`);
        
        return {
            success: true,
            status: response.status,
            data: response.data,
            message: 'Control-M ejecutado exitosamente desde cliente'
        };

    } catch (error) {
        console.error('Error ejecutando Control-M desde cliente:', error.message);
        
        return {
            success: false,
            error: error.message,
            status: error.response?.status || 'N/A',
            message: 'Error ejecutando Control-M desde cliente'
        };
    }
}

/**
 * Función completa: llamar a la API principal y ejecutar Control-M
 * @param {string} apiUrl - URL de tu API principal
 * @param {Object} requestData - Datos para enviar a la API principal
 * @returns {Promise<Object>} Resultado completo del proceso
 */
async function processWithControlM(apiUrl, requestData) {
    try {
        console.log('Llamando a la API principal...');
        
        // Llamar a la API principal
        const response = await axios.post(apiUrl, requestData, {
            headers: {
                'Content-Type': 'application/json'
            }
        });

        if (!response.data.success) {
            throw new Error('La API principal falló');
        }

        console.log('API principal ejecutada exitosamente');
        console.log('Ejecutando Control-M desde cliente...');

        // Ejecutar Control-M desde el cliente
        const controlMResult = await executeControlMFromClient(response.data.controlMInfo);

        return {
            apiResponse: response.data,
            controlMResult: controlMResult,
            success: true,
            message: 'Proceso completo ejecutado exitosamente'
        };

    } catch (error) {
        console.error('Error en el proceso completo:', error.message);
        
        return {
            success: false,
            error: error.message,
            message: 'Error en el proceso completo'
        };
    }
}

// Ejemplo de uso
async function example() {
    const apiUrl = 'https://tu-url-railway.up.railway.app/save-json';
    const requestData = {
        ambiente: 'DEV',
        token: 'tu-bearer-token',
        filename: 'ejemplo-cliente',
        jsonData: {
            jobType: 'Job',
            application: 'MiApp',
            subApplication: 'SubApp'
        }
    };

    const result = await processWithControlM(apiUrl, requestData);
    console.log('Resultado final:', JSON.stringify(result, null, 2));
}

// Exportar funciones
module.exports = {
    executeControlMFromClient,
    processWithControlM
};

// Ejecutar ejemplo si se llama directamente
if (require.main === module) {
    example();
}

File: client-control-m.py
import requests
import json
from io import BytesIO

class ControlMClient:
    """Cliente para ejecutar Control-M desde Python"""
    
    def __init__(self):
        self.session = requests.Session()
    
    def execute_control_m_from_client(self, control_m_info):
        """
        Ejecutar Control-M desde el cliente usando la información obtenida de la API principal
        
        Args:
            control_m_info (dict): Información obtenida de la API principal
            
        Returns:
            dict: Resultado de la ejecución de Control-M
        """
        try:
            print(f"Ejecutando Control-M desde cliente: {control_m_info['url']}")
            
            # Convertir JSON a string y crear buffer
            json_string = json.dumps(control_m_info['jsonData'], indent=2)
            json_bytes = json_string.encode('utf-8')
            
            # Crear form-data
            files = {
                'definitionsFile': (
                    control_m_info['filename'],
                    BytesIO(json_bytes),
                    'application/json'
                )
            }
            
            # Configurar headers
            headers = control_m_info['headers']
            
            # Realizar la petición POST
            response = self.session.post(
                control_m_info['url'],
                files=files,
                headers=headers,
                timeout=30
            )
            
            print(f"Control-M ejecutado exitosamente. Status: {response.status_code}")
            
            return {
                'success': True,
                'status': response.status_code,
                'data': response.json() if response.headers.get('content-type', '').startswith('application/json') else response.text,
                'message': 'Control-M ejecutado exitosamente desde cliente'
            }
            
        except Exception as error:
            print(f"Error ejecutando Control-M desde cliente: {str(error)}")
            
            return {
                'success': False,
                'error': str(error),
                'status': getattr(error.response, 'status_code', 'N/A') if hasattr(error, 'response') else 'N/A',
                'message': 'Error ejecutando Control-M desde cliente'
            }
    
    def process_with_control_m(self, api_url, request_data):
        """
        Función completa: llamar a la API principal y ejecutar Control-M
        
        Args:
            api_url (str): URL de tu API principal
            request_data (dict): Datos para enviar a la API principal
            
        Returns:
            dict: Resultado completo del proceso
        """
        try:
            print("Llamando a la API principal...")
            
            # Llamar a la API principal
            response = self.session.post(
                api_url,
                json=request_data,
                headers={'Content-Type': 'application/json'},
                timeout=30
            )
            
            if response.status_code != 200:
                raise Exception(f"La API principal falló con status: {response.status_code}")
            
            api_response = response.json()
            
            if not api_response.get('success'):
                raise Exception("La API principal retornó error")
            
            print("API principal ejecutada exitosamente")
            print("Ejecutando Control-M desde cliente...")
            
            # Ejecutar Control-M desde el cliente
            control_m_result = self.execute_control_m_from_client(api_response['controlMInfo'])
            
            return {
                'apiResponse': api_response,
                'controlMResult': control_m_result,
                'success': True,
                'message': 'Proceso completo ejecutado exitosamente'
            }
            
        except Exception as error:
            print(f"Error en el proceso completo: {str(error)}")
            
            return {
                'success': False,
                'error': str(error),
                'message': 'Error en el proceso completo'
            }

# Ejemplo de uso
def main():
    client = ControlMClient()
    
    api_url = 'https://tu-url-railway.up.railway.app/save-json'
    request_data = {
        'ambiente': 'DEV',
        'token': 'tu-bearer-token',
        'filename': 'ejemplo-cliente-python',
        'jsonData': {
            'jobType': 'Job',
            'application': 'MiApp',
            'subApplication': 'SubApp'
        }
    }
    
    result = client.process_with_control_m(api_url, request_data)
    print("Resultado final:", json.dumps(result, indent=2))

if __name__ == "__main__":
    main()

